### PR TITLE
WIP Add dask array image module

### PR DESCRIPTION
**Pull request by cowlicks**
This is an effort to wrap `scipy.ndimage.filters`. The current API is like:
```python
dask.array.image.filter_(ndimage_filter, daskarray, **filter_kwargs)
```

- [ ] convolve
- [x] convolve1d
- [ ] correlate
- [x] correlate1d
- [x] gaussian_filter
- [ ] gaussian_filter1d
- [ ] gaussian_gradient_magnitude
- [ ] gaussian_laplace
- [ ] generic_filter
- [ ] generic_filter1d
- [ ] generic_gradient_magnitude
- [ ] generic_laplace
- [ ] laplace
- [ ] maximum_filter
- [ ] maximum_filter1d
- [ ] median_filter
- [ ] minimum_filter
- [ ] minimum_filter1d
- [ ] percentile_filter
- [ ] prewitt
- [ ] rank_filter
- [ ] sobel
- [ ] uniform_filter
- [ ] uniform_filter1d

### CODE REVIEW

File: dask/array/image.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        # choose a deth function from the ndimage filter name
+        depths = {'gaussian_filter' : gauss_depth,
+                  'gaussian_filter1d': gauss1d_depth,
+                  'correlate1d': correlate1d_depth,
+                  'convolve1d' : correlate1d_depth,}
+
+        return depths[filt.__name__](arr, filt, filter_kwargs)
+
+    return {'depth' : depth(arr, filt, filter_kwargs),
+            'boundary' : boundary(filter_kwargs)}
```
Comment (mrocklin) [inline dask/array/image.py]: Nice.

**Comment by mrocklin** [comment]
I wonder if @stefanv has thoughts here?  

We had an in-person conversation in which proposed that a lot of this work should probably happen in scikit-image rather than in dask.array.  The idea being that they're better suited to manage api changes than we would be; they can track changes between the single-threaded and dask.array version.

I'd also generally be happy to offload this work to that community if they're up for it.  They've tried this in the past but ran into technical difficulties.  I wonder if a dask.array solution to https://github.com/scikit-image/scikit-image/pull/723 would be useful.

**Comment by cowlicks** [comment]
Yeah, this is super fragile to changes in ndimage. I am basically hardcoding the way to calculate the `depth` along each axis for each filter. And I'm getting these calculations by inspecting the ndimage codebase. It would be nice if there were an easier way to get this. But I don't think there is.

There are a lot of comments in that PR, but if I understand it correctly, it requires the user to specify the depth of the padding for each filtering operation. It is basically mapping a function across blocks with user specified boundaries. So we could easily do the same thing like:

```python
def dask.array.image.filter_(filter, daskarr, filter_args, boundary, depth):
    A = dask.array.ghost.ghost(daskarr, boundary=boundary, depth=depth)
    B = A.map_blocks(filter, *filter_args)
    return dask.array.ghost.trim_internal(B, depth)
```
If this is all that scikit-image users want, and judging from the interest in that PR then dask array will be very useful for them.


The problem user probably won't know a priori what the depth needs to be. Which is what most of the effort in this PR would be figuring out.

**Comment by mrocklin** [comment]
> The problem user probably won't know a priori what the depth needs to be. Which is what most of the effort in this PR would be figuring out.

Presumably this is an easier operation for them to do than for us to do.  They probably have the domain experience to make this easy.  I think it might be reasonable to punt on this in dask.array and focus on building a general function like what was proposed in that PR but using our machinery.  Perhaps we include some of this work in a PR to scikit-image as a starter?

**Comment by cowlicks** [comment]
> Perhaps we include some of this work in a PR to scikit-image as a starter?

@mrocklin @stefanv If they don't mind having dask as a dependency I would love too!

**Comment by stefanv** [comment]
We are happy to discuss dask as a dependency, given that it is lightweight and pure Python

**Comment by mrocklin** [comment]
Currently `dask.array` depends on `toolz, psutil, chest, numpy`.  We could remove psutil and chest fairly easily.  Other dask things depend on other more exotic libraries 

```Python
extras_require = {
  'array': ['numpy'],
  'bag': ['dill'],
  'distributed': ['zmq'],
  'dataframe': ['bcolz', 'numpy', 'pandas >= 0.16.0']
}
```

**Comment by cowlicks** [comment]
@mrocklin I think we only need `dask.array`, and `numpy>=1.6.1` is already a requirement of scikit-image. So we just have to figure out how we should distribute different parts of dask separately.

**Comment by mrocklin** [comment]
For PyPI I've started using `dask[array]`, `dask[dataframe]`, etc... Presumably we'll need something like metapackages for conda (though conda installers probably won't care as much about lightweight dependencies.)

**Comment by stefanv** [comment]
Can you add these instructions to the docs?

**Comment by mrocklin** [comment]
Yes, http://dask.readthedocs.org/en/latest/install.html

**Comment by shoyer** [comment]
It works fine to just `pip install dask` if you only require dask array and already require numpy in your downstream project.

**Comment by jakirkham** [comment]
Not trying to necro an old thread nor take away from other development efforts. Just thought it would be relevant for others that stumble across it.

Was interested in wrapping these same SciPy functions to use with Dask. After some discussion with others decided to put this in a separate library. It is called [`dask-ndfilters`]( https://github.com/dask-image/dask-ndfilters ). Have implemented many of the listed filters and do intend to cover all of them.  At this time it includes all N-D filters except some of the more obscure generic filters.

**Comment by mrocklin** [comment]
Thanks for the ressurrection @jakirkham .  Do you have any thoughts on how to direct users to things like this?  When someone wants to use Dask for image processing what should they know?  Should we have a docpage somewhere pointing them to the right resources?  Are there other bits of functionality within Dask today that should migrate to some general dask-image or dask-signals package?  Thoughts on how we can increase visibility of your work?

**Comment by jakirkham** [comment]
Good questions.

> Do you have any thoughts on how to direct users to things like this? When someone wants to use Dask for image processing what should they know? Should we have a docpage somewhere pointing them to the right resources?

Yeah, maybe an additional resources page in Dask's docs linking to external/community packages would be good. Could also add a brief blurb in any sections that seem to have crossover. Have also included docs for that package (mainly API at this point), which could be expanded upon. There seem to be a few other Dask related packages that probably could be highlighted in the same way.

> Are there other bits of functionality within Dask today that should migrate to some general dask-image or dask-signals package?

The only thing that I know of that would overlap ATM is `imread`. However I'm thinking about implementing this in a different way as I'd like to use [PIMS]( https://github.com/soft-matter/pims ) to do the heavy lifting probably based off of this [code]( https://github.com/jakirkham/pims_dask_exp ). The motivation being determining metadata (e.g. shape and dtype) in Dask's `imread` currently seems to result in loading full files, but PIMS is able to avoid this by just parsing the header. Unfortunately PIMS doesn't have a good way to integrate with `imread` due to missing attributes and issues pickling objects. ( https://github.com/soft-matter/pims/pull/254 ) ( https://github.com/soft-matter/pims/issues/265 ) So I think it is just going to be easier to reimplement `imread`. Fortunately the code is basically there. So it should be straightforward to get packaged. Though am open to discussion on this and would love to facilitate collaboration with Dask and PIMS in so much as I'm able.

<br>

Kind of a tangent, but still relevant `dask-ndfilters` lives in an organization called [dask-image]( http://dask-image.github.io/ ). It contains a few repos, which break up [SciPy ndimage]( https://docs.scipy.org/doc/scipy-0.19.0/reference/ndimage.html ) into pieces. Not all of them are done yet, but the plan is to complete them over time. One thing I had been musing on was fusing them into a single package once they are more feature complete. Am planning a similar approach with [dask-signal]( https://github.com/dask-signal ).

**Comment by mrocklin** [comment]
Docpage trying (vainly) to list known dask-related packages: https://github.com/dask/dask/pull/2395

**Pull request closed without merging**